integration/test/test_progress.py

```diff
@@ -0,0 +1,379 @@
+        util.assertNear(self, expected_max, max_prog, epsilon=epsilon,
+                        msg='{}: maximum value'.format(msg))
+        # Assert that the expected deviation from the fit is is no
+        # larger than 1%.
+        rr = max_prog - min_prog
+        self.assertLess(0.0, poly_out[0][0])
+        error = (poly_out[1][0] / len(progress)) ** 0.5 / rr
+        self.assertLess(error, 0.01, msg)
+
+    def sum_progress(self, cpu_wp, df):
+        """Sum of all progress values from a set of cpus
+
+        Args:
+            cpu_wp: A container of CPU index values
+
+            df: Data frame to select columns from
+
+        Returns:
+            The sum of all matching REGION_PROGRESS columns from the
+            data frame
+
+        """
+        result = numpy.zeros(len(df))
+        for cpu in cpu_wp:
+            name = 'REGION_PROGRESS-cpu-{}'.format(cpu)
+            result += df[name]
+        return result
+
+    def test_num_host(self):
+        """Check the number of hosts in the report
+
+        """
+        host_names = self._report.host_names()
+        self.assertEqual(len(host_names), self._num_node)
+
+    def get_cpu_with_progress(self):
+        """Find all CPUs for which region progress is not NAN
+
+        """
+        df = self._trace.get_trace_df()
+        result = set()
+        for cpu in range(self._machine.num_cpu()):
+            name = 'REGION_PROGRESS-cpu-{}'.format(cpu)
+            if len(df[name].dropna()) > 0:
+                result.add(cpu)
+        return result
+
+    def test_num_active_cpu(self):
+        """Test that the correct number of CPUs report progress
+
+        """
+        cpu_with_progress = self.get_cpu_with_progress()
+        err_msg ='CPUs with progress:\n    {}\n'.format(cpu_with_progress)
+        self.assertEqual(self._num_active_cpu, len(cpu_with_progress),
+                         msg=err_msg)
+
+    def check_monotone(self, progress, err_msg):
+        """Check that progress does not decrease over time.
+
+        """
+        progress = progress.dropna()
+        self.assertEqual(progress[0], min(progress), err_msg)
+        self.assertEqual(progress[-1], max(progress), err_msg)
+        self.assertLessEqual(0, min(progress.diff()[1:]), err_msg)
+
+    def test_linear_progress_triad(self):
+        """Test that the triad region with with the post reports progress that
+        is well behaved for all CPUs and also collectively.
+
+        """
+        df = self._trace.get_trace_df()
+        grouped_df = df.groupby('REGION_HASH')
+        cpu_with_progress = self.get_cpu_with_progress()
+        max_progress = 1.0 / self._thread_per_rank
+
+        triad_post_hash = self.get_hash('triad_with_post')
+        triad_post_df = grouped_df.get_group(triad_post_hash)
+        total_progress = self.sum_progress(cpu_with_progress, triad_post_df)
+        err_msg = 'Bad fit for triad total proggress'
+        self.check_progress(triad_post_df['TIME'],
```

Comment:
check_monotone() should also be valid for total progress

Reply:
I will try to bring that back.

Reply:
The problem with the total progress analysis is that I had to select based on the region hash for the entire board (i.e. the synchronous part).  This meant that the end of the lagging ranks progress was cut off by the group_by() selection.  This caused issues with the helper check assertsions.  The sum of progress is a little contrived, and checking each of them is a finer grained check (now selecting on per-cpu region hash).